src/browser/BrowserService.cpp

```diff
@@ -0,0 +1,570 @@
+    //    var n = String.Join("\n    ", names.ToArray<string>());
+    //    if (BrowserSettings::receiveCredentialNotification())
+    //        ShowNotification(QString("%0: %1 is receiving credentials for:\n%2").arg(Id).arg(host).arg(n)));
+    //}
+
+    // Fill the list
+    for (Entry* entry : pwEntries)
+        result << prepareEntry(entry);
+
+    return result;
+}
+
+void BrowserService::addEntry(const QString &, const QString &login, const QString &password, const QString &url, const QString &submitUrl, const QString &realm)
+{
+    if (Group* group = findCreateAddEntryGroup()) {
+        Entry* entry = new Entry();
+        entry->setUuid(Uuid::random());
+        entry->setTitle(QUrl(url).host());
+        entry->setUrl(url);
+        entry->setIcon(KEEPASSBROWSER_DEFAULT_ICON);
+        entry->setUsername(login);
+        entry->setPassword(password);
+        entry->setGroup(group);
+
+        const QString host = QUrl(url).host();
+        const QString submitHost = QUrl(submitUrl).host();
+        BrowserEntryConfig config;
+        config.allow(host);
+        if (!submitHost.isEmpty())
+            config.allow(submitHost);
+        if (!realm.isEmpty())
+            config.setRealm(realm);
+        config.save(entry);
+    }
+}
+
+void BrowserService::updateEntry(const QString &id, const QString &uuid, const QString &login, const QString &password, const QString &url)
+{
+    if (thread()!=QThread::currentThread())
+    {
+        QMetaObject::invokeMethod(this, "updateEntry", Qt::BlockingQueuedConnection,    Q_ARG(const QString&, id),
+                                                                                        Q_ARG(const QString&, uuid),
+                                                                                        Q_ARG(const QString&, login),
+                                                                                        Q_ARG(const QString&, password),
+                                                                                        Q_ARG(const QString&, url));
+    }
+
+    if (DatabaseWidget* dbWidget = m_dbTabWidget->currentDatabaseWidget())
+        if (Database* db = dbWidget->database())
+            if (Entry* entry = db->resolveEntry(Uuid::fromHex(uuid))) {
+                QString u = entry->username();
+                if (u != login || entry->password() != password) {
+                    //ShowNotification(QString("%0: You have an entry change prompt waiting, click to activate").arg(requestId));
+                    if (BrowserSettings::alwaysAllowUpdate()
+                        || QMessageBox::warning(0, tr("KeePassXC: Update Entry"),
+                                                tr("Do you want to update the information in %1 - %2?")
+                                                .arg(QUrl(url).host().toHtmlEscaped()).arg(u.toHtmlEscaped()),
+                                                QMessageBox::Yes|QMessageBox::No) == QMessageBox::Yes ) {
```

Comment:
You shouldn't use HTML escaping here (I know, we have code like that, but still...)
Explicitly set the message box text format to plain text instead.